File: 0061-rotate-list/0061-rotate-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null || k == 0) {
            return head;
        }
        int length = 1;
        ListNode current = head;
        while (current.next != null) {
            current = current.next;
            length++;
        }
        int newHeadPosition = length - k % length;
        if (newHeadPosition == length) {
            return head;
        }
        current.next = head; // Make the list circular
        current = head;
        for (int i = 0; i < newHeadPosition - 1; i++) {
            current = current.next;
        }
        head = current.next;
        current.next = null;
        return head;
    }
}

File: 0061-rotate-list/0061-rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    void rotateByone(ListNode* &head){
        ListNode* temp=head;
        if (head == NULL || head->next == NULL) {
             return;
        }
        while(temp->next->next!=NULL){
            temp=temp->next;
        }
        temp->next->next=head;
        head=temp->next;
        temp->next=NULL;
    }
    ListNode* rotateRight(ListNode* head, int k) {
        int n=0;
        ListNode* curr=head;
        if (head == NULL || head->next == NULL) {
             return head;
        }
        while(curr!=NULL){
            n++;
            curr=curr->next;
        }

        k=k%n;
        if(k==0){
            return head;
        }
        while(k>0){
           rotateByone(head);
           k--; 
        }
        return head;

    }
};